### ISSUE TITLE
Standardize error + flash responses

### ISSUE THREAD

**Issue by ojeytonwilliams** [freeCodeCamp/freeCodeCamp#50184]
Currently the api-server sends errors messages back to the client in at least two ways. Either as a JSON response or as a query param. The former take the following shape:

```
{
  "message": "flash.updated-preferences",
  "type": "success",
  "variables": { "are": "optional" }
}
```

The query params are

```
{
  "messages": [ // elements have above shape ]
}
```

however, Fastify's default error responses are like this:

```
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": "body/profileUI must have required property 'isLocked'"
}
```

which is useful for debugging purposes.

However, we cannot send both JSON responses simultaneously. Also, we cannot merge the two, since we'd either lose the user friendly message (flash) or the detailed error message.

## Issues to resolve

What do we want long term? Some combined object? For example:

```
{
  "statusCode": 400,
  "error": "Bad Request",
  "errorMessage": "body/profileUI must have required property 'isLocked'",
  "message": "flash.updated-preferences",
  "type": "success",
  "variables": { "are": "optional" }
}
```

What do we want to do in the interim? Defer fixing this until after the first release? Respond with the flash message data for now?

**Comment by ShaunSHamilton** [comment]
It might be nice to have a `createResponse(code, message)` function we had to use everywhere.

```typescript
function createResponse(code: number, message?: string): ReplyObject {
  // Attach code, use message or default
  return { ... };
}
```

---

After thinking about it, the flash system seems like something that should be entirely handled (logicked) on the client's side. The client should deal with defaults, and what kind of flash to show based on the error code and/or response message.

**Comment by ShaunSHamilton** [comment]
> Otherwise, that's pretty much what we already have.

Well, yes. For the most part. However, I do not think we should be sending `flash.<KIND>` in `message`. We should send the raw info, and the client should be responsible for the "user readable" message.

**Comment by ojeytonwilliams** [comment]
Maybe, yeah. It depends if the client always has the info it needs (when you take the error code into account). I suspect that's not always the case, though.

If it _is_ the case, I'm all for letting the client handle that logic.

**Comment by moT01** [comment]
I agree that, ideally, the client should handle what is displayed based on the response. The client should know what the user was attempting to do - i.e. Click "Save Privacy Settings" -> error -> display "Something went wrong trying to save your privacy settings"

I think this would simplify the API. We should bring it up in the next meeting.

**Comment by Viksil11** [comment]
No nude pedo

**Comment by ShaunSHamilton** [comment]
Thinking about this more due to the Microsoft stuff where a single request can fail in so many ways, I think it would be nice to take a chapter out of IDLs, and have our error enums be something like:

```typescript
// utils/shareable-errors.ts
// This will be defined by the API, but can be consumed by any clients

enum MsError {
  InvalidLinkDomain = "The URL should match the 'learn.microsoft.com' domain",
  UncompletedModule = "One or more modules for this trophy have not been completed at 'learn.microsoft.com'"
}
```

Then, during the build, the `shareable-errors.ts` would be transpiled into:

```json
// Translatable error messages
{
  "errors": [
    {
      "code": 0,
      "name": "InvalidLinkDomain",
      "msg": "The URL should match the..."
    },
    {
      "code": 1,
      "name": "UncompletedModule",
      "msg": "One or more..."
    }
  ]
}
```

**Comment by ShaunSHamilton** [comment]
I think we should be going for as little information as required.
> I.e. we can drop the 'type', because the client can see the status code

I would drop almost everything that is not default (Fastify/prisma), because the client should already be aware of the variables (request data).

**Comment by moT01** [comment]
Perhaps that's the way to go. In theory, I like it - but just kind of thinking out loud...

A possible response:

```
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": "body/profileUI must have required property 'isLocked'"
}
```

So then on the client we would have a lookup object or something?

```
{
  "body/profileUI must have required property 'isLocked'": "flash.is-locked-err"
}
```

Seems like an extra step - but not terrible I guess. Do we know that the client will have access to all the variables? We have this endpoint for the MS trophies:

```
const msUser = await MsUsername.findOne({
  where: { userId: user.id }
});

const { msUsername } = msUser;
const msApiRes = await fetch(msTrophyApiUrl);
const msTrophyJson = await msApiRes.json();

if (!msApiRes.ok || msTrophyJson.awardType !== 'Trophy') {
  return res.status(403).json({
    type: 'error',
    message: 'flash.ms.trophy.err-3',
    variables: {
      msUsername
    }
  });
}
```

The client should have their MsUsername - but if it's different than the one found in the database, it could help us figure out what is going wrong?

**Comment by ShaunSHamilton** [comment]
> but if it's different than the one found in the database, it could help us figure out what is going wrong?

Good point. We should definitely be returning the relevant information. I suppose the current `{message, variables}` format is ok to keep.

We should not be telling the client what to do, though (`flash...`). We could just have `message: "4xx.ms.trophy-3"`:

```ts
void res.statusCode(403); // Implies `type: "error"`
return {
  message: "4xx.ms.trophy-3" // Notice the `4xx` implies `.err`
  variables: {
    msUsername
  }
}
```

```json
{
  "4xx": {
    "ms": {
      "trophy-3": "Our usual format {{msUsername}}"
    }
  }
}
```

Then, we just need to publish our `.json` file with map to possible responses. Also, it would probably be better if we published some helper/package to use our _custom_ format.

**Comment by ojeytonwilliams** [comment]
We shouldn't forget the human readable messages. `4xx.ms.trophy-3` doesn't mean anything to me.

**Comment by ShaunSHamilton** [comment]
> We shouldn't forget the human readable messages. `4xx.ms.trophy-3` doesn't mean anything to me.

We would still include whatever default comes with, but without duplicating info in two files, we either:
a) Have `4xx.ms.trophy-3` and map it to a `responses.json` file
b) Have fancy tooling that can automatically build all our responses `{ message: "Our usual format {{msUsername}}" }` into a `responses.json` file for clients to consume.

The only reason for this rigmarole is for i18n.

**Comment by ojeytonwilliams** [comment]
Sorry, I wasn't talking about the i18n. That's all reasonable, but I think we'll still want a message that we, as developers, can understand.

So, in this hypothetical, we'd have

```js
{
  .. stuff
  "message": "userName missing from Microsoft transcript response"
}
```

That way we can see what the problem is without having to dig into the json.

**Comment by ShaunSHamilton** [comment]
Right,
> but without duplicating info in two files...

Because then we would need to have:

```ts
// script.ts
return { message: "userName missing ...", mark: "4xx.ms.trophy-3" }
```

```json
// responses.json
{
  "...trophy-3": "userName missing ..."
}
```

**Comment by ojeytonwilliams** [comment]
I doubt it would be duplicated, because we'd want to send different info to developers and to users.

**Comment by moT01** [comment]
I don't think I fully understand what some of these suggestions will look like - I still kinda like [my earlier suggestion.](https://github.com/freeCodeCamp/freeCodeCamp/issues/50184#issuecomment-1740873166) Having the `flash` property in the response is more like a suggestion from the server - it's still up to the client to determine what to do with it. Also, it should be a pretty easy refactor on the client.

> ``void res.statusCode(403); // Implies type: "error"``

In most (maybe all) cases, this is accurate - but there may be an instance where we do want an 'info' style message when there's a 4xx error. The client could still override it and display the info style - but that seems like some potentially unnecessary extra logic on the client. It just seems simple to send a `flash` to client and then the client can just display it.

I'm not sure if this needs to come into the discussion, but we do have endpoints that redirect. Could we make those more restful, by maybe adding a `redirect` property, and let the client do the redirecting? Or maybe we want to try and get rid of those or something.

Whatever we decide, I would vote for whatever is the simplest, easiest to understand, and easiest to develop on.

**Comment by moT01** [comment]
This feels post-mvp. Moved it to there, let me know if I shouldn't have.

**Comment by KhiraHamdi** [comment]
I found this issue still exists in the current codebase. 
In src/components/Flash/index.tsx line 20, the workaround 
mapping 'error' to 'danger' is still in place. The TODO 
comment confirms this needs standardization at the API level. 
Happy to help test any fix for this.
